src/main/java/com/thealgorithms/strings/AhoCorasick.java

```diff
@@ -0,0 +1,181 @@
+            root.child.get(rc).suffix_link = root; // Set suffix link to the root
+        }
+
+        while (!q.isEmpty()) {
+            Node currentState = q.poll(); // Get the current node for processing
+
+            // Iterate through child nodes of the current node
+            for (char cc : currentState.child.keySet()) {
+                Node currentChild = currentState.child.get(cc); // Get the child node
+                Node parentSuffix = currentState.suffix_link; // Get the parent's suffix link
+
+                // Calculate the suffix link for the child based on parent's suffix link
+                while (!parentSuffix.child.containsKey(cc) && parentSuffix != root) {
+                    parentSuffix = parentSuffix.suffix_link;
+                }
+
+                // Set the calculated suffix link or default to root
+                if (parentSuffix.child.containsKey(cc)) {
+                    currentChild.suffix_link = parentSuffix.child.get(cc);
+                } else {
+                    currentChild.suffix_link = root;
+                }
+
+                q.add(currentChild); // Add the child node to the queue for further processing
+            }
+
+            // Establish output links for nodes to efficiently identify patterns within patterns
+            if (currentState.suffix_link.pattern_ind >= 0) {
+                currentState.output_link = currentState.suffix_link;
+            } else {
+                currentState.output_link = currentState.suffix_link.output_link;
+            }
+        }
+    }
+
+    // Searches for patterns in the input text and records their positions
+    public void searchIn(String text) {
+        Node parent = root; // Start searching from the root node
+        for (int i = 0; i < text.length(); i++) {
+            char ch = text.charAt(i); // Get the current character in the text
+
+            // Check if the current node has a child for the current character
+            if (parent.child.containsKey(ch)) {
+                parent = parent.child.get(ch); // Update the current node to the child node
+
+                // If the current node represents a pattern, record its position in res
+                if (parent.pattern_ind > -1) {
+                    res.get(parent.pattern_ind).add(i);
+                }
+
+                Node output_link = parent.output_link;
+                // Follow output links to find and record positions of other patterns
+                while (output_link != null) {
+                    res.get(output_link.pattern_ind).add(i);
+                    output_link = output_link.output_link;
+                }
+            } else {
+                // If no child node exists for the character, backtrack using suffix links
+                while (parent != root && !parent.child.containsKey(ch)) {
+                    parent = parent.suffix_link;
+                }
+                if (parent.child.containsKey(ch)) {

```

Comment:
In the existing tests, this condition if never `false`.

Reply:
Got it. Will add a test case regarding this.

Reply:
Added test-case, regarding this with few more test-cases.
It is also done now.